### PR TITLE
[WIP] UnitedDeployment supports AdvancedStatefulSet: introduce an interface for each kind of subset

### PR DESCRIPTION
**Pull request by wu8685**
### Ⅰ. Describe what this PR does

Introduce an interface in SubsetControl of UnitedDeployment. UD can support each kind of subset by implementing it.

### CODE REVIEW

File: pkg/controller/uniteddeployment/adapter/statefulset_adapter.go [hunk 1]
```diff
@@ -0,0 +1,260 @@
...
+	return &appsv1.StatefulSet{}
+}
+
+// NewResourceListObject creates a empty StatefulSetList object.
+func (a *StatefulSetAdapter) NewResourceListObject() runtime.Object {
+	return &appsv1.StatefulSetList{}
+}
+
+// GetObjectMeta returns the ObjectMeta of the subset of StatefulSet.
+func (a *StatefulSetAdapter) GetObjectMeta(obj metav1.Object) *metav1.ObjectMeta {
```
Comment (Fei-Guo) [inline pkg/controller/uniteddeployment/adapter/statefulset_adapter.go]: Can we try not expose this as interface?

File: pkg/controller/uniteddeployment/adapter/statefulset_adapter.go [hunk 2]
```diff
@@ -0,0 +1,260 @@
...
+}
+
+// GetSubsetFailure returns the failure information of the subset.
+// StatefulSet has no condition.
+func (a *StatefulSetAdapter) GetSubsetFailure() *string {
+	return nil
+}
+
+// ConvertToResourceList converts StatefulSetList object to StatefulSet array.
+func (a *StatefulSetAdapter) ConvertToResourceList(obj runtime.Object) []metav1.Object {
```
Comment (Fei-Guo) [inline pkg/controller/uniteddeployment/adapter/statefulset_adapter.go]: I doubt if we need to expose this as interface method. Can we move this to subset_control.go as a private method?

File: pkg/controller/uniteddeployment/adapter/statefulset_adapter.go [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+		return nil
+	}
+
+	// If RollingUpdate, work around for issue https://github.com/kubernetes/kubernetes/issues/67250
+	return a.deleteStuckPods(set, revision, partition)
+}
+
+// IsExpected checks the subset is the expected revision or not.
+// The revision label can tell the current subset revision.
+func (a *StatefulSetAdapter) IsExpected(ud *alpha1.UnitedDeployment, obj metav1.Object, revision string) bool {
```
Comment (Fei-Guo) [inline pkg/controller/uniteddeployment/adapter/statefulset_adapter.go]: ud is not used.

File: pkg/controller/uniteddeployment/adapter/adapter.go [hunk 4]
```diff
@@ -0,0 +1,31 @@
+package adapter
```
[the remark this answers is not in the document]
Comment (wu8685) [reply pkg/controller/uniteddeployment/adapter/adapter.go]: done

File: pkg/controller/uniteddeployment/adapter/adapter_util.go [hunk 5]
```diff
@@ -0,0 +1,81 @@
+package adapter
```
[the remark this answers is not in the document]
Comment (wu8685) [reply pkg/controller/uniteddeployment/adapter/adapter_util.go]: done

File: pkg/controller/uniteddeployment/adapter/statefulset_adapter.go [hunk 6]
```diff
@@ -0,0 +1,260 @@
+package adapter
```
[the remark this answers is not in the document]
Comment (wu8685) [reply pkg/controller/uniteddeployment/adapter/statefulset_adapter.go]: done

File: pkg/controller/util/statefulset_utils.go [hunk 7]
```diff
@@ -0,0 +1,29 @@
+package util
```
[the remark this answers is not in the document]
Comment (wu8685) [reply pkg/controller/util/statefulset_utils.go]: done

File: pkg/controller/uniteddeployment/adapter/statefulset_adapter.go (hunk 1 above)
Comment (wu8685) [reply pkg/controller/uniteddeployment/adapter/statefulset_adapter.go]: done

File: pkg/controller/uniteddeployment/adapter/statefulset_adapter.go (hunk 2 above)
Comment (wu8685) [reply pkg/controller/uniteddeployment/adapter/statefulset_adapter.go]: Use reflect, instead.
Please take a look in `subset_control.go` L59.

File: pkg/controller/uniteddeployment/adapter/statefulset_adapter.go (hunk 3 above)
Comment (wu8685) [reply pkg/controller/uniteddeployment/adapter/statefulset_adapter.go]: Used in `uniteddeployment_update.go` L53.

**Pull request merged**